test/distributed/_shard/checkpoint/test_file_system_checkpoint.py

```diff
@@ -0,0 +1,506 @@
+    else:
+        gather_list = None
+
+    # FIXME is a rank allowed to not have any data?
+    with torch.no_grad():
+        # XXX we can fastpath this to torch.cat if max_rank_size == rank_sizes[rank]
+        data = torch.empty(max_rank_size, device=self.local_shards()[0].tensor.device)
+        for shard in self.local_shards():
+            for placement in local_shards_placement:
+                if placement[0] == shard.metadata:
+                    src = shard.tensor.flatten()
+                    data[placement[1]: placement[1] + src.numel()].copy_(src)
+                    break
+
+    dist.gather(
+        tensor=data,
+        gather_list=gather_list,
+        dst=dst,
+        group=self._process_group,
+    )
+    if rank != dst:
+        return
+    if out is None:
+        raise ValueError("`out` Tensor must be provided on dst rank!")
+
+    full_size = self.metadata().size
+    dims = len(full_size)
+
+
+    for shard_idx, shard_md in enumerate(self.metadata().shards_metadata):
+        placement = shard_placement[shard_idx]
+        tensor = gather_list[placement[0]]
+        tensor = tensor[placement[1] : placement[1] + shard_size(shard_md)]
+        tensor = tensor.view(shard_md.shard_sizes)
+
+        out_narrow_view = out
+        for dim in range(dims):
+            out_narrow_view = out_narrow_view.narrow(
+                dim,
+                shard_md.shard_offsets[dim],
+                shard_md.shard_sizes[dim],
+            )
+
+        out_narrow_view.copy_(tensor)
+
+
+
+def assert_state_dict_equal(
+    self: TestCase,
+    state_dict_1: Dict[str, torch.Tensor],
+    state_dict_2: Dict[str, torch.Tensor],
+) -> bool:
+    self.assertEqual(
+        len(state_dict_1), len(state_dict_2), "state_dict must be the same size"
+    )
+    self.assertEqual(
+        set(state_dict_1.keys()),
+        set(state_dict_2.keys()),
+        "state_dict keys do not match",
+    )
+
+    for key, value_1 in state_dict_1.items():
+        value_2 = state_dict_2[key]
+        if isinstance(value_1, torch.Tensor):
+            self.assertTrue(
+                torch.equal(value_1, value_2), f"Key {key}'s tensor does not match"
+            )
+        elif isinstance(value_1, ShardedTensor):
+            for local_shard_1, local_shard_2 in zip(
+                value_1.local_shards(), value_2.local_shards()
+            ):
+                self.assertTrue(
+                    torch.equal(local_shard_1.tensor, local_shard_1.tensor),
+                    f"Key {key}'s shard does not match",
+                )
+
+    return True
+
+
+class MyTestModule(torch.nn.Module):
+    def __init__(self) -> None:
+        super().__init__()
+        self.linear_1 = torch.nn.Linear(5, 5)
+        self.linear_2 = torch.nn.Linear(5, 1)
+        self.emb = torch.nn.EmbeddingBag(5, 10)
+
+
+# The ShardedModels are borrowed from test/distributed/_sharded_tensor/test_sharded_tensor.py
+class MyShardedModel3(torch.nn.Module):
```

Comment:
nit: Maybe move these to a common location and reuse? Ex: torch/testing/_internal/distributed/_shard/sharded_tensor/_test_st_common.py

Reply:
👍